### PR TITLE
Type inf improvements

### PR DESCRIPTION
**Pull request by ZacLN**
Adds some rudimentary inference for elements of arrays. 

For type inference by use (i.e. where we infer the type of a variable by the argument slot it sits in in a function call) adds handling for `if` blocks so that the uncertainty of whether a branch is hit is accounted for.

Also adds some internal (as yet unused) functions for reasoning about the specific method being used at a call site. The intention is to use this in combination with function return type cached by SymbolServer to improve our inference. [Note - there are currently two different pieces of code that iterate across function methods (`check_call` and `infer_type_by_use`) and use of the functions referenced above would add a third. These should be combined/rationalised.]

For every PR, please check the following:
- [x] End-user documentation check. If this PR requires end-user documentation in the Julia VS Code extension docs, please add that at https://github.com/julia-vscode/docs.
- [x] Changelog mention. If this PR should be mentioned in the CHANGELOG for the Julia VS Code extension, please open a PR against https://github.com/julia-vscode/julia-vscode/blob/master/CHANGELOG.md with those changes.

### CODE REVIEW

File: src/type_inf.jl [hunk 1]
```diff
@@ -32,42 +32,51 @@ end
...
+        end
+    else
+        if CSTParser.is_func_call(rhs)
+            callname = CSTParser.get_name(rhs)
+            if isidentifier(callname)
+                resolve_ref(callname, scope, state)
+                if hasref(callname)
+                    rb = get_root_method(refof(callname), state.server)
+                    if (rb isa Binding && (CoreTypes.isdatatype(rb.type) || rb.val isa SymbolServer.DataTypeStore)) || rb isa SymbolServer.DataTypeStore
+                        settype!(binding, rb)
```
Comment (pfitzseb) [inline src/type_inf.jl:50]: Can we end up with abstract types here? If so, is that a problem?

**Review by pfitzseb — DISMISSED** [review]
Awesome!

File: src/type_inf.jl (hunk 1 above)
Comment (ZacLN) [reply src/type_inf.jl:50]: This is all slightly dodgy - it assumes datatype methods _are_ constructors that return an instance. Clearly there's no requirement for this but would for the vast majority of cases be true

Comment (pfitzseb) [reply src/type_inf.jl:50]: Yeah, that seems like a good assumption to make.

File: src/type_inf.jl [hunk 2]
```diff
@@ -57,6 +57,8 @@ function infer_type_assignment_rhs(binding, state, scope)
             settype!(binding, CoreTypes.Float64)
         elseif CSTParser.isstringliteral(rhs)
             settype!(binding, CoreTypes.String)
+        elseif headof(rhs) === :TRUE || headof(rhs) === :false
```
Comment (pfitzseb) [inline src/type_inf.jl:60]: Why's one head `:TRUE` and the other `:false`? (C/S)hould we make those consistent?

**Comment by ZacLN** [comment]
@pfitzseb this is ready to go. I've added some other code since last reviewed but this is largely unused as of yet.

**Comment by davidanthoff** [comment]
Needs a merge conflict resolution.

**Review by davidanthoff — APPROVED** [review]

**Pull request merged**